cpp/src/arrow/dataset/file_skyhook.cc

```diff
@@ -0,0 +1,280 @@
+#include "arrow/dataset/file_skyhook.h"
+
+#include <mutex>
+
+#include "arrow/api.h"
+#include "arrow/compute/exec/expression.h"
+#include "arrow/dataset/dataset_internal.h"
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/file_ipc.h"
+#include "arrow/dataset/file_parquet.h"
+#include "arrow/filesystem/filesystem.h"
+#include "arrow/filesystem/path_util.h"
+#include "arrow/filesystem/util_internal.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/compression.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/logging.h"
+#include "parquet/arrow/reader.h"
+#include "parquet/file_reader.h"
+
+#include <flatbuffers/flatbuffers.h>
+
+#include "generated/ScanRequest_generated.h"
+
+namespace arrow {
+
+namespace flatbuf = org::apache::arrow::flatbuf;
+
+namespace dataset {
+
+namespace connection {
+
+std::mutex connection_mutex;
+
+RadosConnection::~RadosConnection() { Shutdown(); }
+
+Status RadosConnection::Connect() {
+  if (connected) {
+    return Status::OK();
+  }
+
+  // Locks the mutex. Only one thread can pass here at a time.
+  // Another thread handled the connection already.
+  std::unique_lock<std::mutex> lock(connection_mutex);
+  if (connected) {
+    return Status::OK();
+  }
+  connected = true;
+
+  if (rados->init2(ctx.user_name.c_str(), ctx.cluster_name.c_str(), 0))
+    return Status::Invalid("librados::init2 returned non-zero exit code.");
+
+  if (rados->conf_read_file(ctx.ceph_config_path.c_str()))
+    return Status::Invalid("librados::conf_read_file returned non-zero exit code.");
+
+  if (rados->connect())
+    return Status::Invalid("librados::connect returned non-zero exit code.");
+
+  if (rados->ioctx_create(ctx.data_pool.c_str(), ioCtx))
+    return Status::Invalid("librados::ioctx_create returned non-zero exit code.");
+
+  return Status::OK();
+}
+
+Status RadosConnection::Shutdown() {
+  rados->shutdown();
+  return Status::OK();
+}
+
+}  // namespace connection
+
+/// \brief A ScanTask to scan a file fragment in Skyhook format.
+class SkyhookScanTask : public ScanTask {
+ public:
+  SkyhookScanTask(std::shared_ptr<ScanOptions> options,
+                  std::shared_ptr<Fragment> fragment, FileSource source,
+                  std::shared_ptr<SkyhookDirectObjectAccess> doa, int fragment_format)
+      : ScanTask(std::move(options), std::move(fragment)),
+        source_(std::move(source)),
+        doa_(std::move(doa)),
+        fragment_format_(fragment_format) {}
+
+  Result<RecordBatchIterator> Execute() override {
+    struct stat st {};
+    ARROW_RETURN_NOT_OK(doa_->Stat(source_.path(), st));
+
+    ceph::bufferlist request;
+    ARROW_RETURN_NOT_OK(
+        SerializeScanRequest(options_, fragment_format_, st.st_size, request));
+
+    ceph::bufferlist result;
+    ARROW_RETURN_NOT_OK(doa_->Exec(st.st_ino, "scan_op", request, result));
+
+    RecordBatchVector batches;
+    ARROW_RETURN_NOT_OK(DeserializeTable(batches, result, !options_->use_threads));
```

Comment:
Why `!options_->use_threads`?

Reply:
I did this, because earlier arrow didn't support nested threading very well, and what used to happen was threaded decompression caused the execution to enter an deadlocked state.

Reply:
Looks like the decompression API also needs to be futurized for the async scanner to work. Since, it is not yet futurized,  looks like we need to do this check to avoid nested threading.